### PR TITLE
feat(query): Adding quota parameters for individual query limits and further reporting.

### PR DESCRIPTION
**Pull request by kvpetrov**
Adding quota parameters for individual query limits and further reporting.

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 
Currently, there are some limits established for query processing, however, these limits are either defaults in the scala code or properties of FiloDB runtime. 

**New behavior :**
We want to be able to pass these limits as part of the query utilizing PlannerParams

**BREAKING CHANGES**
Protos defining PlannerParams are redefined in a manner which is not backwards compatible. This should be ok as our scala pojos normally are not backwards compatible. 

**Other information**:

### CODE REVIEW

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (alextheimer) [inline core/src/main/scala/filodb.core/query/QueryContext.scala:46]: Are these separate "warning" thresholds needed? Is it sufficient to warn at X% of the limit?

If the separate thresholds are needed, a `Threshold(warning, limit)` class is another option to consider (this way we wouldn't need something like `apply(warnDefaults)`).

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (alextheimer) [inline core/src/main/scala/filodb.core/query/QueryContext.scala:68]: I'm not sure where we use this specific `apply`, but it might construct `PlannerParams` with some warning limits less than the enforced limits.

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala
Comment (alextheimer) [inline query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:81]: Where will `warnQuota` be used?

**Review by alextheimer — DISMISSED** [review]
Gotta love those `ExecPlan::printTree` comparison tests 😛
Approved-- please consider some of the comments I left before you merge.

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (amolnayak311) [inline query/src/main/scala/filodb/query/ProtoConverters.scala:202]: anti pattern for use of ``map``,  it should not cause side effects, either use ``isDefined`` or ``foreach``

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (amolnayak311) [inline query/src/main/scala/filodb/query/ProtoConverters.scala:244]: It's in inappropriate to add ``warnDefaults`` here. This converter should convert given proto to equivalent case class. Misusing this method will introduce bugs hard to troubleshoot. ``toProto`` and ``fromProto`` should not accept parameters and alter the behavior. . 

Have another one to convert case class to proto and write a spec similar to how it's written for others.

**Review by amolnayak311 — CHANGES_REQUESTED** [review]
Please see changes and incorporate the changes requested. Thanks for putting this together.

**Comment by amolnayak311** [comment]
Please update the description to give details if the PR and commit as per [these](https://github.com/filodb/FiloDB/blob/develop/CONTRIBUTING.md) guidelines.

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:46]: I'm only copying whatever was already in the existing properties. This PR essentially groups limits in an object and allows to have two sets of quotas/limits: (1) enforced (2) warn and allows to pass these parameters from the QueryService but it DOES NOT change the existing behavior.

Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:46]: The warning piece is fundamental for the PR. The reason we do not enforce the limits (we have very high limits) is because it's impossible for us to enforce them. There would be a number of users whose queries will stop to work and we simply cannot let it happen (imagine alerts or recording rules), instead, we need to:
A) establish quotas that we believe are resonable on WARN level
B) find out all the users/cases where the limits are breached
C) establish SPECIFIC HIGH limits for recording rules/alerts/ some API queries
D) enforce reasonable limits for EVERYBODY but the specific cases above
E) notify the users with specific high limits to let them know that their RR/alerts will be disabled in X days if they don't take care of it.

At this stage I think it's reasonable to have more flexibility in defining what the warn and enforce levels are. The ratio is hard to come up by looking at the crystal ball. Maybe it's 5x maybe 2. This we will figure out as part of an experiment around steps A/B/C.

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:68]: this is expected, isn't it? the warning limits are smaller than enforced, ie at 2MB I warn you, at 10MB I won't let your query run.

File: query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/exec/BinaryJoinExec.scala:81]: pls, see the comment above, I will also update the description in the PR

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:25]: can we do it later, in the second PR, that PR will do nothing but renaming the object?

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:54]: pls see comment above, will update the description in the PR as well.
In short the warn most likely in the first phase will go to logs. It would be used by operators to see what the reasonable limits are.

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (kvpetrov) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:53]: I think we can move them to QueryContext too as part of a different PR, originally the parameters were int PlannerParams.

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:202]: not sure I understood this comment. queryOrigin is an option. I set queryOrigin on the proto builder if and only if option is not None. Why side effects here?

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:244]: need to talk to you about this one I guess. the problem is that it seems like protos don't have default values but scala case classes do. So, suppose you have a proto that does not have a property defined but you cannot NOT to create a corresponding scala object since its properties are not optional. So, either I define proto properties as required or?..  I noticed that most of your protos have the fields defined as optional and I believe you did it for a reason, correct?

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:183]: there is one defined:
Line 243.
implicit class IndividualQuotaFromProtoConverter(giq: GrpcMultiPartitionQueryService.IndividualQuota) {

File: core/src/main/scala/filodb.core/query/QueryContext.scala
Comment (amolnayak311) [reply core/src/main/scala/filodb.core/query/QueryContext.scala:25]: I leave it to you and @vishramachandran the later we do more we will have to work on preserving backward compatibility.

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:183]: It just converts the proto to case class (``fromProto``), we need a ``toProto`` to convert the case class to gRPC object. Anywhere we deal with converting to and from proto we should use this implicit conversions rather than copying the object building logic as we have in this PR

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:202]: Your ``map`` is setting something in an object (side effect), by design this should simply transform and give a new option transforming the original one. As I mentioned, use ``foreach`` or ``isDefined`` to set this proto field.

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:183]: got you, will add toProto (only fromProto is defined now)

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (amolnayak311) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:244]: Correct but we never add parameters to ``toProto`` and ``fromProto``. The goal is to convert whatever the case class has (default or explicitly set values) to proto and back.

File: query/src/main/scala/filodb/query/ProtoConverters.scala
Comment (kvpetrov) [reply query/src/main/scala/filodb/query/ProtoConverters.scala:202]: will do

**Review by amolnayak311 — APPROVED** [review]
Thanks @kvpetrov for the changes. LGTM. I have retriggered the PRB which was failing.

**Pull request merged**
